src/NHibernate/Async/Collection/Generic/PersistentGenericSet.cs

```diff
@@ -54,7 +54,7 @@ public override async Task<bool> EqualsSnapshotAsync(ICollectionPersister persis
 		{
 			cancellationToken.ThrowIfCancellationRequested();
 			var elementType = persister.ElementType;
-			var snapshot = (ISetSnapshot<T>)GetSnapshot();
+			var snapshot = (SetSnapShot<T>)GetSnapshot();
```

Comment:
Can't we add `GetSnapshotTyped()` returning `SetSnapShot<T>` and avoid typing in all methods

Reply:
Added.

Reply:
That's not exactly what I meant. Typing is still there - it's just moved to separate method. I meant extract content of `GetSnapshot` to `GetSnapshotTyped`

Reply:
I do not get what or how @bahusoid. We have `GetSnapshot() => session.PersistenceContext.GetSnapshot(this)`, we anyway have to put the typing somewhere, unless rewriting a lot of things to be generic, but that would not fit in this PR scope.

Reply:
We have `PersistentGenericSet.GetSnapshot`. Why can't content of it be extracted to separate method:
```C#
public override object GetSnapshot(ICollectionPersister persister)
{
	return GetSnapshotSet(persister);
}

private SetSnapShot<T> GetSnapshotSet(ICollectionPersister persister)
{
	var clonedSet = new SetSnapShot<T>(WrappedSet.Count);
	var enumerable = from object current in WrappedSet
					select persister.ElementType.DeepCopy(current, persister.Factory);
	foreach (var copied in enumerable)
	{
		clonedSet.Add((T) copied);
	}

	return clonedSet;
}
```

And use it inside this class to avoid typing only inside `PersistentGenericSet`

Reply:
It wrecks the logic. The snapshot must be taken from the persistence context, not from the `WrappedSet`. Otherwise the collection will never be considered dirty. It seems to me you are mixing up `GetSnapshot()` with `GetSnapshot(ICollectionPersister persister)`. Although they share the same name, their purpose is not the same. The former get the snapshot representing the collection state as it was in the database upon loading, while the later extract a snapshot from the current collection state.

Reply:
Yeah.. Sorry  my mistake...

Reply:
Async regen is missed for `GetSnapshotTyped`. Or just kill commit adding it as it's not very useful in the end..

Reply:
Reverted commit.